egs/wsj/s5/steps/nnet3/multilingual/allocate_multilingual_examples.py

```diff
@@ -187,121 +141,79 @@ def process_multilingual_egs(args):
-            raise Exception("Error opening file {0}".format(o))
-        if w is None:
-            raise Exception("Error opening file {0}".format(w))
-        if scp_per_archive_file is None:
-            raise Exception("Error opening file {0}".format(scp_per_archive_file))
-
-        for job in range(args.num_jobs):
-            scp = ("{0}/temp/{1}scp.{2}.{3}".format(args.egs_dir, args.egs_prefix,
-                                                    job + 1, archive + 1))
-            with open(scp, "r") as scpfile:
-                for line in scpfile:
-                    scp_line = line.splitlines()[0].split()
-                    print("{0} {1}".format(scp_line[0], scp_line[1]),
-                          file=scp_per_archive_file)
-                    print("{0} output-{1}".format(scp_line[0], scp_line[2]),
-                          file=o)
-                    print("{0} {1}".format(
-                            scp_line[0],
-                            lang2weight[int(scp_line[2])]),
-                          file=w)
-            os.remove(scp)
-
-        for (lang_id, start_eg_line, num_egs) in all_egs[num_archives * job + archive]:
-            this_ranges.append((lang_id, start_eg_line, num_egs))
-
-        # write egs.ranges.*.txt
-        for (lang_id, start_eg_line, num_egs) in this_ranges:
-            print("{0} {1} {2}".format(lang_id, start_eg_line, num_egs), file=f)
-
-        f.close()
-        o.close()
-        w.close()
-        scp_per_archive_file.close()
-    logger.info("finished generating {0}*.scp, {0}output.*.ark "
+    if not os.path.exists(os.path.join(args.egs_dir, 'info')):
+        os.makedirs(os.path.join(args.egs_dir, 'info'))
+
+    with open("{0}/info/{1}num_tasks".format(args.egs_dir, args.egs_prefix), "w") as fh:
+        print("{0}".format(num_langs), file=fh)
+
+    # Total number of egs in all languages
+    tot_num_egs = sum(lang_to_num_examples[i] for i in range(num_langs))
+    if args.num_archives is not None:
+        num_archives = args.num_archives
+    else:
+        num_archives = tot_num_egs // args.samples_per_iter + 1
+
+
+    with open("{0}/info/{1}num_archives".format(args.egs_dir, args.egs_prefix), "w") as fh:
+        print("{0}".format(num_archives), file=fh)
+
+    egs_per_archive = tot_num_egs // num_archives
+    avg_block_size = args.block_size
+
+    # This is the number of robin rounds we do for generating each output scp (in
+    # each round we read one block from each input lang and write it to the current
+    # output scp file; please see the main 'for' loop below):
+    num_rounds = egs_per_archive // (avg_block_size * num_langs)
+
+    # The block size for each lang is calculated based on its size (and average block size).
+    # Note that we can't use the same block size for all input languages; otherwise
+    # we might run out of egs for some of the languages.
+    lang_to_block_size = [int(lang_to_num_examples[i] / tot_num_egs * avg_block_size * num_langs) + 1
```

Comment:
I am not sure allowing less than block size is ok. If it is less than minibatch size, those examples will be thrown out during training. So no training will be done for those tasks.

Reply:
This number can be very small, if lang_to_num_examples[i] / tot_num_egs = 0.01, then lang_to_block_size is very small (e.g. 1)